ros_ws/src/teleoperation/src/keyboard_controller.cpp

```diff
@@ -1,73 +1,126 @@
 #include "keyboard_controller.h"
 
+double clamp(double value, double lower, double upper)
+{
+    return min(upper, max(value, lower));
+}
+
+double map(double value, double in_lower, double in_upper, double out_lower, double out_upper)
+{
+    return out_lower + (out_upper - out_lower) * (value - in_lower) / (in_upper - in_lower);
+}
+
 KeyboardController::KeyboardController()
 {
-    this->drive_parameters_publisher = this->node_handle.advertise<drive_msgs::drive_param>(TOPIC_DRIVE_PARAMETERS, 1);
+    this->m_drive_parameters_publisher =
+        this->m_node_handle.advertise<drive_msgs::drive_param>(TOPIC_DRIVE_PARAMETERS, 1);
+    this->createWindow();
 }
 
-void KeyboardController::keyboardLoop()
+KeyboardController::~KeyboardController()
 {
-    std::cout << "Listening to keyboard..." << std::endl;
-    std::cout << "========================" << std::endl;
+    SDL_FreeSurface(this->m_window);
+    SDL_Quit();
+}
 
-    double velocity = 0;
-    double angle = 0;
-    while (ros::ok())
+void KeyboardController::createWindow()
+{
+    if (SDL_Init(SDL_INIT_VIDEO) < 0)
     {
-        auto key = static_cast<Keycode>(this->getKeyboardCharacter());
+        throw std::runtime_error("Could not initialize SDL");
+    }
+    SDL_EnableKeyRepeat(SDL_DEFAULT_REPEAT_DELAY, SDL_DEFAULT_REPEAT_INTERVAL);
+    SDL_WM_SetCaption("Keyboard teleoperation - Use WASD keys", NULL);
+    this->m_window = SDL_SetVideoMode(500, 150, 0, 0);
+}
 
-        switch (key)
+void KeyboardController::pollKeyboardEvents()
+{
+    SDL_Event event;
+    while (SDL_PollEvent(&event))
+    {
+        if (event.type == SDL_KEYUP || event.type == SDL_KEYDOWN)
         {
-            case Keycode::W:
-                velocity += 1;
-                break;
-            case Keycode::S:
-                velocity -= 1;
-                break;
-            case Keycode::A:
-                angle += 1;
-                break;
-            case Keycode::D:
-                angle -= 1;
-                break;
+            for (int i = 0; i < KEY_COUNT; i++)
+            {
+                if ((uint16_t)event.key.keysym.sym == (int)KEY_CODES[i])
```

Comment:
Here auto-casting might results in a undesired result. Either cast both values to the same type or declare `KEY_CODES` the same type as `event.key.keysym.sym`.